### PR TITLE
Extensions inherit the availability of the extended type

### PR DESCRIPTION
**Pull request by xymus**
Automatically apply the availability of the extended type to its extensions. As such, there is no need to repeat the availability attribute on extensions for them to have the same availability.

### CODE REVIEW

**Comment by xymus** [comment]
@swift-ci Please smoke test

**Comment by xymus** [comment]
@swift-ci Please test source compatibility

**Comment by xymus** [comment]
@swift-ci Please smoke test

**Comment by xymus** [comment]
@swift-ci Please test source compatibility

**Comment by xymus** [comment]
@swift-ci Please smoke test

**Comment by xymus** [comment]
@swift-ci Please test source compatibility

File: lib/Sema/TypeCheckAvailability.cpp [hunk 1]
```diff
@@ -656,6 +656,18 @@ TypeChecker::overApproximateAvailabilityAtLocation(SourceLoc loc,
     }
 
     DC = D->getDeclContext();
+
+    // Look for the availability of the extended type of an ExtensionDecl.
+    if (auto extension = dyn_cast<ExtensionDecl>(DC)) {
+      if (auto extended = extension->getExtendedNominal()) {
+        Optional<AvailabilityContext> Info =
+          AvailabilityInference::annotatedAvailableRange(extended, Context);
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckAvailability.cpp]: Concerned that this won't handle this case:

```swift
@available(macOS 10.15, *)
struct Outer {
  struct Inner {}
}

extension Outer.Inner {
  func test() {}
}
```

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

**Pull request closed without merging**